pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/BaseForwardIndexBasedIndexHandler.java

```diff
@@ -0,0 +1,93 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.segment.index.loader;
+
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.SegmentMetadata;
+import org.apache.pinot.segment.spi.creator.IndexCreatorProvider;
+import org.apache.pinot.segment.spi.store.ColumnIndexType;
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Base class for all of the {@link IndexHandler} classes which need to build their indexes from the forward index.
+ * This class handles temporarily rebuilding the forward index if the forward index does not exist and cleaning up
+ * the forward index once all handlers have completed via overriding the postUpdateIndicesCleanup() method.
+ */
+public abstract class BaseForwardIndexBasedIndexHandler implements IndexHandler {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BaseForwardIndexBasedIndexHandler.class);
+
+  protected final SegmentMetadata _segmentMetadata;
+  protected final IndexLoadingConfig _indexLoadingConfig;
+  protected final Set<String> _tmpForwardIndexColumns;
+
+  public BaseForwardIndexBasedIndexHandler(SegmentMetadata segmentMetadata, IndexLoadingConfig indexLoadingConfig) {
+    _segmentMetadata = segmentMetadata;
+    _indexLoadingConfig = indexLoadingConfig;
+    _tmpForwardIndexColumns = new HashSet<>();
+  }
+
+  @Override
+  public void postUpdateIndicesCleanup(SegmentDirectory.Writer segmentWriter)
+      throws Exception {
+    // Delete the forward index for columns which have it disabled. Perform this as a post-processing step after all
+    // IndexHandlers have updated their indexes as some of them need to temporarily create a forward index to
+    // generate other indexes off of.
+    for (String column : _tmpForwardIndexColumns) {
+      segmentWriter.removeIndex(column, ColumnIndexType.FORWARD_INDEX);
+    }
```

Comment:
Do we have tests for the scenario that check for temporarily rebuilding forward index to generate some other index and then temp index gets deleted ?

Reply:
I have added tests which require a forward index but it is disabled. I don't validate that we created a forward index and then deleted it, but I do validate that the index I expected to create now exists. Is this sufficient?

Reply:
> I have added tests which require a forward index but it is disabled.

Yes, this is good and queries will fail. 

> but I do validate that the index I expected to create now exists.

this is fine too. 

>  don't validate that we created a forward index and then deleted it

Was wondering if we can verify **both creation (which you are already doing I guess) + cleanup** of temp fwd index to ensure there is no lingering temp fwd index after preprocessor finishes.

Reply:
> Was wondering if we can verify both creation (which you are already doing I guess) + cleanup of temp fwd index to ensure there is no lingering temp fwd index after preprocessor finishes.

I do validate that the forward index doesn't exist if it is disabled. So this is covered. e.g. say we enabled range index on forward index disabled column, I validate range index exists and forward index doesn't exist.

Reply:
Cool. Sounds good.